certificate-authority/store/mongodb/bulkWriter.go

```diff
@@ -0,0 +1,235 @@
+	flushTimeout  time.Duration
+	logger        log.Logger
+
+	done    chan struct{}
+	trigger chan bool
+
+	mutex  sync.Mutex
+	models map[string]*store.SigningRecord
+	wg     sync.WaitGroup
+}
+
+func newBulkWriter(col *mongo.Collection, documentLimit uint16, throttleTime time.Duration, flushTimeout time.Duration, logger log.Logger) *bulkWriter {
+	r := &bulkWriter{
+		col:           col,
+		documentLimit: documentLimit,
+		throttleTime:  throttleTime,
+		flushTimeout:  flushTimeout,
+		done:          make(chan struct{}),
+		trigger:       make(chan bool, 1),
+		logger:        logger,
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		r.run()
+	}()
+	return r
+}
+
+func toSigningRecordFilter(signingRecord *store.SigningRecord) bson.M {
+	res := bson.M{"_id": signingRecord.GetId()}
+	return res
+}
+
+func getSigningRecordCreationDate(defaultTime time.Time, signingRecord *store.SigningRecord) int64 {
+	ret := defaultTime.UTC().UnixNano()
+	if signingRecord.GetCredential().GetDate() > 0 && signingRecord.GetCredential().GetDate() < ret {
+		ret = signingRecord.GetCredential().GetDate()
+	}
+	return ret
+}
+
+func setValueByDate(key, datePath string, dateOperator string, date int64, value interface{}) bson.M {
+	return bson.M{
+		"$set": bson.M{
+			key: bson.M{
+				"$ifNull": bson.A{
+					bson.M{
+						"$cond": bson.M{
+							"if": bson.M{
+								dateOperator: bson.A{"$" + datePath, date},
+							},
+							"then": value,
+							"else": "$" + key,
+						},
+					}, value,
+				},
+			},
+		},
+	}
+}
+
+func updateSigningRecord(signingRecord *store.SigningRecord) []bson.M {
+	creationDate := signingRecord.GetCreationDate()
+	if creationDate == 0 {
+		creationDate = getSigningRecordCreationDate(time.Now(), signingRecord)
+	}
+	ret := []bson.M{
+		{"$set": bson.M{
+			"_id":               signingRecord.GetId(),
+			store.CommonNameKey: signingRecord.GetCommonName(),
+			store.OwnerKey:      signingRecord.GetOwner(),
+			store.PublicKeyKey:  signingRecord.GetPublicKey(),
+		}},
+	}
+	ret = append(ret, setValueByDate(store.CreationDateKey, store.CreationDateKey, "$gt", creationDate, creationDate))
+	if signingRecord.GetCredential() != nil {
+		ret = append(ret, setValueByDate(store.CredentialKey, store.CredentialKey+"."+store.DateKey, "$lt", signingRecord.GetCredential().GetDate(), signingRecord.GetCredential()))
+	}
+	return ret
+}
+
+func convertSigningRecordToWriteModel(signingRecord *store.SigningRecord) mongo.WriteModel {
+	return mongo.NewUpdateOneModel().SetFilter(toSigningRecordFilter(signingRecord)).SetUpdate(updateSigningRecord(signingRecord)).SetUpsert(true)
+}
+
+func mergeLatestUpdateSigningRecord(toUpdate *store.SigningRecord, latest *store.SigningRecord) *store.SigningRecord {
+	if toUpdate == nil {
+		return latest
+	}
+	if latest.GetCredential().GetDate() > toUpdate.GetCredential().GetDate() {
+		toUpdate.Credential = latest.GetCredential()
+	}
+	if latest.GetCreationDate() < toUpdate.GetCreationDate() {
```

Comment:
lines 115-124 are not used according to Sonar, can't we add a test case that verifies that this works properly?